### PR TITLE
PrecursorAction.C code cleanup

### LINKED ISSUE #139: PrecursorAction.C code cleanup
PrecursorAction.C contains many repeating if-else conditionals and a code structure that is tedious to read. In contrast, it is much easier to follow the logic in INSAction.C in the Navier-Stokes module in MOOSE

This issue can be closed when we've applied the DRY (Don't Repeat Yourself) principle and properly "functionalized" the code in PrecursorAction.C.

### PR DESCRIPTION
**Pull request by smpark7**
Closes #139 

After reviewing PrecursorAction.C,
- I introduced the ```setVarNameAndBlock``` function to remove some repeating code. Further attempts to "DRY" would only complicate the code and affect human readability.
- I refactored the action functions and separated them based on their "physics"/"purpose" (e.g. ```kernelAct()``` -> ```addPrecursorSource()``` + ```addPrecursorDecay()``` + etc.). The main ```act()``` function now holds more if-else logic explicitly for better readability
- Removed some unnecessary vertical whitespaces

There is no change in how PrecursorAction evaluates precursor concentration => All existing tests should pass

### CODE REVIEW

**Review by katyhuff — CHANGES_REQUESTED** [review]
Great job @smpark7 !  Some minor comments included in this review, but nothing major.

File: src/actions/PrecursorAction.C [hunk 1]
```diff
@@ -76,7 +68,7 @@ validParams<PrecursorAction>()
       "",
       "An optional suffix string that can be helpful to avoid object name crashing.");
   params.addParam<std::vector<SubdomainName>>("kernel_block",
-                                              "Kernel bock can be different from block.");
+                                              "Kernel block can be different from block.");
```
[the remark this answers is not in the document]
Comment (robfairh) [reply src/actions/PrecursorAction.C:71]: I am not sure. Precursor block maybe?

File: src/actions/PrecursorAction.C [hunk 2]
```diff
@@ -140,166 +132,148 @@ PrecursorAction::act()
...
+        addTimeDerivative(var_name);
+
+      addPrecursorSource(op, var_name);
+      addPrecursorDecay(op, var_name);
+    }
 
     // dg kernels
     if (_current_task == "add_dg_kernel")
-      dgKernelAct(var_name);
+    {
```
[the remark this answers is not in the document]
Comment (smpark7) [reply src/actions/PrecursorAction.C:145]: Yes that makes sense. I'll change all of the ```_current_task``` ifs to elseifs and remove unnecessary braces

File: src/actions/PrecursorAction.C [hunk 3]
```diff
@@ -140,166 +132,148 @@ PrecursorAction::act()
...
+      addDGAdvection(var_name);
+    }
 
     // bcs
     if (_current_task == "add_bc")
-      bcAct(var_name);
+    {
+      addOutflowBC(var_name);
+
+      if (getParam<bool>("loop_precs"))
```
[the remark this answers is not in the document]
Comment (smpark7) [reply src/actions/PrecursorAction.C:154]: Sounds good. I'll make that change. I'll also have to edit the test input files but I think it's worth the hassle

File: src/actions/PrecursorAction.C [hunk 4]
```diff
@@ -140,166 +132,148 @@ PrecursorAction::act()
...
+        addInflowBC(var_name);
+    }
 
     // ics
     if (_current_task == "add_ic" && !getParam<bool>("init_from_file"))
-      icAct(var_name);
+      // Set up initial conditions for precursor conc
```
[the remark this answers is not in the document]
Comment (smpark7) [reply src/actions/PrecursorAction.C:160]: Got it. I'll rename it into something like ```addInitialConditions```

File: src/actions/PrecursorAction.C [hunk 5]
```diff
@@ -140,166 +132,148 @@ PrecursorAction::act()
...
+      addOutletPostprocessor(var_name);
+      addInletPostprocessor(var_name);
+    }
 
     // transfers
     if (_current_task == "add_transfer" && getParam<bool>("loop_precs") &&
         !getParam<bool>("is_loopapp"))
-      transferAct(var_name);
+    {
+      // Set up MultiAppTransfer to simulate precursor looped flow into and
```
[the remark this answers is not in the document]
Comment (smpark7) [reply src/actions/PrecursorAction.C:176]: Yes it is. I'll make them consistent by naming both MultiAppTransfer. Naming one of them FlowTransfer seems unnecessary

File: src/actions/PrecursorAction.C [hunk 6]
```diff
@@ -140,166 +132,148 @@ PrecursorAction::act()
...
   }
 
   // Add outflow rate postprocessor for Navier-Stokes velocities in the main
   // app if precursors are looped
-  if (_current_task == "add_postprocessor" && isParamValid("uvel") &&
-      getParam<bool>("loop_precs") && !getParam<bool>("is_loopapp"))
-    addOutflowPostprocessor();
+  if (_current_task == "add_postprocessor" && getParam<bool>("loop_precs") &&
+      isParamValid("uvel") && !getParam<bool>("is_loopapp"))
+    addSaltOutflowPostprocessor();
```
[the remark this answers is not in the document]
Comment (smpark7) [reply src/actions/PrecursorAction.C:186]: That's a really good point. I had not considered that there could be other reactor designs with moving precursors (other than perhaps the certain pebble bed reactors).

**Comment by smpark7** [comment]
Thank you for your review @katyhuff 

I've made the following changes based on your review comments:
- Appropriate changes from ```if``` to ```else if``` involving the ```_current_task ==``` loop
- Renaming functions and variables to make them more intuitive and removing explicit references to salt
- Improving/Adding docstrings for functions and variables
- Made most functions in ```PrecursorAction``` non-virtual because it isn't meant to be a base class

File: src/actions/PrecursorAction.C (hunk 5 above)
[the remark this answers is not in the document]
Comment (katyhuff) [reply src/actions/PrecursorAction.C:176]: Thanks!

File: src/actions/PrecursorAction.C [hunk 7]
```diff
@@ -140,166 +132,148 @@ PrecursorAction::act()
...
-    std::string kernel_name = "ScalarTransportTimeDerivative_" + var_name + "_" + _object_suffix;
-    _problem->addKernel("ScalarTransportTimeDerivative", kernel_name, params);
-  }
+  std::string kernel_name = "ScalarTransportTimeDerivative_" + var_name + "_" + _object_suffix;
+  _problem->addKernel("ScalarTransportTimeDerivative", kernel_name, params);
 }
 
 void
-PrecursorAction::dgKernelAct(const std::string & var_name)
+PrecursorAction::addDGAdvection(const std::string & var_name)
```
[the remark this answers is not in the document]
Comment (katyhuff) [reply src/actions/PrecursorAction.C:231]: This seems to have been handled, yes? If so, please 'resolve' conversation. Thanks.

File: src/actions/PrecursorAction.C [hunk 8]
```diff
@@ -140,166 +132,148 @@ PrecursorAction::act()
...
+    params.set<FunctionName>("vel_y_func") = getParam<FunctionName>("v_func");
+    params.set<FunctionName>("vel_z_func") = getParam<FunctionName>("w_func");
+    std::string kernel_name = "DGFunctionConvection_" + var_name + "_" + _object_suffix;
+    _problem->addDGKernel("DGFunctionConvection", kernel_name, params);
   }
 }
 
 void
-PrecursorAction::bcAct(const std::string & var_name)
+PrecursorAction::addOutflowBC(const std::string & var_name)
```
[the remark this answers is not in the document]
Comment (katyhuff) [reply src/actions/PrecursorAction.C:272]: This seems like it's probably been handled. If so, please 'resolve' conversation.

File: src/actions/PrecursorAction.C (hunk 7 above)
[the remark this answers is not in the document]
Comment (katyhuff) [reply src/actions/PrecursorAction.C:231]: nvm, found it. Resolved! Thanks!

**Review by katyhuff — APPROVED** [review]
Looks good to me! Any remaining issues @smpark7 or @robfairh ?  I'd say this is ready to merge.

**Comment by smpark7** [comment]
Thank you @katyhuff! @robfairh feel free to merge it unless you spot another issue

**Review by robfairh — APPROVED** [review]
Katy's review was very thorough!

**Pull request merged**